File: implementations/python/happ/adapters/entra_oidc_pkce.py
from __future__ import annotations

import os
import secrets
import time
import urllib.parse
from dataclasses import dataclass
from typing import Any, Dict, Optional

import requests

from happ.crypto.jws import jws_verify_rs256
from happ.util import b64url_encode, sha256_b64url


_CACHE_TTL_SECONDS = 300
_DISCOVERY_CACHE: Dict[str, tuple[float, Dict[str, Any]]] = {}
_JWKS_CACHE: Dict[str, tuple[float, Dict[str, Any]]] = {}


@dataclass(frozen=True)
class EntraOidcConfig:
    tenant_id: str
    client_id: str
    redirect_uri: str
    scope: str = "openid profile email"
    authorize_base: str = "https://login.microsoftonline.com"
    token_base: str = "https://login.microsoftonline.com"


def pkce_create_verifier() -> str:
    # 43-128 chars. Use 64 bytes urlsafe.
    return b64url_encode(secrets.token_bytes(64))


def pkce_challenge(verifier: str) -> str:
    return sha256_b64url(verifier.encode("ascii"))


def build_authorize_url(
    cfg: EntraOidcConfig,
    state: str,
    nonce: str,
    code_verifier: str,
    prompt: Optional[str] = None,
    login_hint: Optional[str] = None,
    domain_hint: Optional[str] = None,
    extra_params: Optional[Dict[str, str]] = None,
) -> str:
    endpoint = f"{cfg.authorize_base}/{cfg.tenant_id}/oauth2/v2.0/authorize"
    params = {
        "client_id": cfg.client_id,
        "response_type": "code",
        "redirect_uri": cfg.redirect_uri,
        "response_mode": "query",
        "scope": cfg.scope,
        "state": state,
        "nonce": nonce,
        "code_challenge": pkce_challenge(code_verifier),
        "code_challenge_method": "S256",
    }
    if prompt:
        params["prompt"] = prompt
    if login_hint:
        params["login_hint"] = login_hint
    if domain_hint:
        params["domain_hint"] = domain_hint
    if extra_params:
        params.update(extra_params)
    return endpoint + "?" + urllib.parse.urlencode(params)


def exchange_code_for_tokens(
    cfg: EntraOidcConfig,
    code: str,
    code_verifier: str,
    client_secret: Optional[str] = None,
    timeout_seconds: int = 15,
) -> Dict[str, Any]:
    """
    Exchange an auth code for tokens.

    NOTE: This performs a live network call to Entra. It will fail in offline environments.
    """
    endpoint = f"{cfg.token_base}/{cfg.tenant_id}/oauth2/v2.0/token"
    data = {
        "client_id": cfg.client_id,
        "grant_type": "authorization_code",
        "code": code,
        "redirect_uri": cfg.redirect_uri,
        "code_verifier": code_verifier,
    }
    if client_secret:
        data["client_secret"] = client_secret
    resp = requests.post(endpoint, data=data, timeout=timeout_seconds)
    resp.raise_for_status()
    return resp.json()


def _cached_get(cache: Dict[str, tuple[float, Dict[str, Any]]], key: str) -> Optional[Dict[str, Any]]:
    entry = cache.get(key)
    if entry is None:
        return None
    expires_at, value = entry
    if expires_at <= time.time():
        cache.pop(key, None)
        return None
    return value


def _cached_put(cache: Dict[str, tuple[float, Dict[str, Any]]], key: str, value: Dict[str, Any]) -> Dict[str, Any]:
    cache[key] = (time.time() + _CACHE_TTL_SECONDS, value)
    return value


def fetch_openid_configuration(cfg: EntraOidcConfig, timeout_seconds: int = 10) -> Dict[str, Any]:
    url = os.environ.get(
        "HAPP_ENTRA_OPENID_CONFIG_URL",
        f"{cfg.authorize_base}/{cfg.tenant_id}/v2.0/.well-known/openid-configuration",
    )
    cached = _cached_get(_DISCOVERY_CACHE, url)
    if cached is not None:
        return cached

    resp = requests.get(url, timeout=timeout_seconds)
    resp.raise_for_status()
    metadata = resp.json()
    if not isinstance(metadata, dict):
        raise ValueError("OpenID configuration response must be a JSON object")
    if not metadata.get("issuer") or not metadata.get("jwks_uri"):
        raise ValueError("OpenID configuration missing issuer or jwks_uri")
    return _cached_put(_DISCOVERY_CACHE, url, metadata)


def fetch_jwks(jwks_url: str, timeout_seconds: int = 10) -> Dict[str, Any]:
    override = os.environ.get("HAPP_ENTRA_JWKS_URL", "").strip()
    url = override or jwks_url
    cached = _cached_get(_JWKS_CACHE, url)
    if cached is not None:
        return cached

    resp = requests.get(url, timeout=timeout_seconds)
    resp.raise_for_status()
    jwks = resp.json()
    if not isinstance(jwks, dict):
        raise ValueError("JWKS response must be a JSON object")
    if not isinstance(jwks.get("keys"), list):
        raise ValueError("JWKS missing keys array")
    return _cached_put(_JWKS_CACHE, url, jwks)


def verify_id_token(
    *,
    id_token: str,
    jwks: Dict[str, Any],
    expected_issuer: str,
    expected_audience: str,
    expected_nonce: str,
    clock_skew_seconds: int = 30,
) -> Dict[str, Any]:
    payload = jws_verify_rs256(id_token, jwks, expected_aud=expected_audience)
    now = int(time.time())

    iss = payload.get("iss")
    if iss != expected_issuer:
        raise ValueError("ID token issuer mismatch")
    nonce = payload.get("nonce")
    if nonce != expected_nonce:
        raise ValueError("ID token nonce mismatch")

    for field_name in ("exp", "iat", "nbf"):
        value = payload.get(field_name)
        try:
            payload[field_name] = int(value)
        except (TypeError, ValueError) as exc:
            raise ValueError(f"ID token {field_name} missing or invalid") from exc

    if payload["exp"] < now - clock_skew_seconds:
        raise ValueError("ID token expired")
    if payload["iat"] > now + clock_skew_seconds:
        raise ValueError("ID token issued in the future")
    if payload["nbf"] > now + clock_skew_seconds:
        raise ValueError("ID token not yet valid")
    if not payload.get("tid") or not payload.get("oid"):
        raise ValueError("ID token missing tid or oid")
    return payload


def env_config() -> EntraOidcConfig:
    tenant_id = os.environ.get("HAPP_ENTRA_TENANT_ID", "common")
    client_id = os.environ.get("HAPP_ENTRA_CLIENT_ID", "")
    redirect_uri = os.environ.get("HAPP_ENTRA_REDIRECT_URI", "http://127.0.0.1:8787/entra/callback")
    scope = os.environ.get("HAPP_ENTRA_SCOPE", "openid profile email")
    return EntraOidcConfig(tenant_id=tenant_id, client_id=client_id, redirect_uri=redirect_uri, scope=scope)

File: implementations/rust/crates/happ-provider/src/adapters/entra_oidc_pkce.rs
use std::collections::HashMap;
use std::sync::Arc;

use async_trait::async_trait;
use chrono::{TimeZone, Utc};
use rand::{distributions::Alphanumeric, Rng};
use serde::Deserialize;
use serde_json::Value;
use url::Url;

use happ_core::{
    types::{
        IdentityAssurance, IdentityBinding, IdentityEvidence, IdentityIdp, IdentityPolicy,
        IdentityRequirements, IdentitySubject,
    },
    HappError, HappResult,
};
use happ_crypto::util::sha256_base64url_nopad;

use crate::adapters::{IdentityAdapter, IdentityAdapterOutcome};
use crate::provider::{OidcTempState, Provider, Session};

#[derive(Clone, Debug)]
pub struct EntraOidcConfig {
    /// Tenant id or special values: common, organizations, consumers
    pub tenant: String,
    pub client_id: String,
    pub client_secret: Option<String>,
    pub redirect_base: String, // e.g. http://127.0.0.1:8787
}

impl EntraOidcConfig {
    pub fn issuer(&self) -> String {
        format!("https://login.microsoftonline.com/{}/v2.0", self.tenant)
    }

    pub fn authorize_endpoint(&self) -> String {
        format!(
            "https://login.microsoftonline.com/{}/oauth2/v2.0/authorize",
            self.tenant
        )
    }

    pub fn token_endpoint(&self) -> String {
        format!(
            "https://login.microsoftonline.com/{}/oauth2/v2.0/token",
            self.tenant
        )
    }

    pub fn jwks_endpoint(&self) -> String {
        format!(
            "https://login.microsoftonline.com/{}/discovery/v2.0/keys",
            self.tenant
        )
    }

    pub fn redirect_uri(&self) -> String {
        format!(
            "{}/identity/entra_oidc/callback",
            self.redirect_base.trim_end_matches('/')
        )
    }
}

/// Microsoft Entra ID OIDC adapter (Authorization Code + PKCE).
///
/// This is an interoperable identity-binding adapter that yields `tid+oid` subject binding.
pub struct EntraOidcPkceAdapter {
    pub cfg: EntraOidcConfig,
    http: reqwest::Client,
}

impl EntraOidcPkceAdapter {
    pub fn new(cfg: EntraOidcConfig) -> Self {
        EntraOidcPkceAdapter {
            cfg,
            http: reqwest::Client::new(),
        }
    }

    fn random_string(len: usize) -> String {
        rand::thread_rng()
            .sample_iter(&Alphanumeric)
            .take(len)
            .map(char::from)
            .collect()
    }

    fn build_pkce(verifier: &str) -> String {
        // S256: BASE64URL-ENCODE(SHA256(ASCII(code_verifier)))
        sha256_base64url_nopad(verifier.as_bytes())
    }

    async fn fetch_jwks(&self) -> HappResult<Value> {
        let jwks_url = self.cfg.jwks_endpoint();
        let v = self
            .http
            .get(jwks_url)
            .send()
            .await
            .map_err(|e| HappError::Io(e.to_string()))?
            .json::<Value>()
            .await
            .map_err(|e| HappError::Io(e.to_string()))?;
        Ok(v)
    }

    fn find_session_by_state(provider: &Provider, state: &str) -> Option<String> {
        for s in provider.list_sessions() {
            if let Some(oidc) = &s.oidc {
                if oidc.csrf == state && oidc.scheme == "entra_oidc" {
                    return Some(s.session_id.clone());
                }
            }
        }
        None
    }

    fn policy_from_req(req: &IdentityRequirements) -> IdentityPolicy {
        req.policy.clone().unwrap_or(IdentityPolicy {
            require_verified: true,
            require_embedded_evidence: false,
            allowed_tenants: vec![],
            max_id_age_seconds: None,
            require_mfa: false,
            required_auth_contexts: vec![],
            allow_asserted: false,
        })
    }
}

#[derive(Debug, Deserialize)]
struct TokenResponse {
    #[serde(rename = "access_token")]
    _access_token: Option<String>,
    id_token: Option<String>,
    #[serde(rename = "refresh_token")]
    _refresh_token: Option<String>,
    #[serde(rename = "expires_in")]
    _expires_in: Option<i64>,
    #[serde(rename = "scope")]
    _scope: Option<String>,
    #[serde(rename = "token_type")]
    _token_type: Option<String>,
}

#[async_trait]
impl IdentityAdapter for EntraOidcPkceAdapter {
    fn scheme(&self) -> &'static str {
        "entra_oidc"
    }

    async fn begin(
        &self,
        provider: Arc<Provider>,
        session: &Session,
        _req: &IdentityRequirements,
        _base_url: &str,
    ) -> HappResult<IdentityAdapterOutcome> {
        let csrf = Self::random_string(32);
        let nonce = Self::random_string(32);
        let pkce_verifier = Self::random_string(64);
        let pkce_challenge = Self::build_pkce(&pkce_verifier);

        provider.set_oidc_state(
            &session.session_id,
            OidcTempState {
                scheme: "entra_oidc".to_string(),
                csrf: csrf.clone(),
                nonce: nonce.clone(),
                pkce_verifier: pkce_verifier.clone(),
            },
        )?;

        let mut url = Url::parse(&self.cfg.authorize_endpoint())
            .map_err(|e| HappError::Invalid(e.to_string()))?;

        {
            let mut qp = url.query_pairs_mut();
            qp.append_pair("client_id", &self.cfg.client_id);
            qp.append_pair("response_type", "code");
            qp.append_pair("redirect_uri", &self.cfg.redirect_uri());
            qp.append_pair("response_mode", "query");
            qp.append_pair("scope", "openid profile email");
            qp.append_pair("state", &csrf);
            qp.append_pair("nonce", &nonce);
            qp.append_pair("code_challenge", &pkce_challenge);
            qp.append_pair("code_challenge_method", "S256");
            // Optional: enforce re-auth
            qp.append_pair("prompt", "select_account");
        }

        // NOTE: In a hardened implementation you may inject Entra auth context requirements
        // via claims challenge or other supported mechanisms. This reference keeps it minimal.
        // The RP expresses policy in req.policy.required_auth_contexts.

        Ok(IdentityAdapterOutcome::Redirect {
            url: url.to_string(),
        })
    }

    async fn handle_callback(
        &self,
        provider: Arc<Provider>,
        _session_id: &str,
        query: &HashMap<String, String>,
        _base_url: &str,
    ) -> HappResult<IdentityAdapterOutcome> {
        let code = query
            .get("code")
            .ok_or_else(|| HappError::Invalid("missing code".to_string()))?
            .clone();
        let state = query
            .get("state")
            .ok_or_else(|| HappError::Invalid("missing state".to_string()))?
            .clone();

        let sid = Self::find_session_by_state(&provider, &state)
            .ok_or_else(|| HappError::Invalid("unknown oidc state".to_string()))?;

        let session = provider
            .get_session(&sid)
            .ok_or_else(|| HappError::NotFound(format!("session {sid} not found")))?;

        let oidc = provider
            .take_oidc_state(&sid)?
            .ok_or_else(|| HappError::Invalid("missing oidc state".to_string()))?;

        if oidc.csrf != state {
            return Err(HappError::Unauthorized("state mismatch".to_string()));
        }

        let idreq = session
            .requirements
            .identity
            .clone()
            .ok_or_else(|| HappError::Invalid("identity not requested".to_string()))?;

        let policy = Self::policy_from_req(&idreq);

        // Exchange auth code for tokens
        let mut form: Vec<(&str, String)> = vec![
            ("client_id", self.cfg.client_id.clone()),
            ("grant_type", "authorization_code".to_string()),
            ("code", code.clone()),
            ("redirect_uri", self.cfg.redirect_uri()),
            ("code_verifier", oidc.pkce_verifier.clone()),
            ("scope", "openid profile email".to_string()),
        ];
        if let Some(secret) = &self.cfg.client_secret {
            form.push(("client_secret", secret.clone()));
        }

        let token_resp = self
            .http
            .post(self.cfg.token_endpoint())
            .form(&form)
            .send()
            .await
            .map_err(|e| HappError::Io(e.to_string()))?
            .json::<TokenResponse>()
            .await
            .map_err(|e| HappError::Io(e.to_string()))?;

        let id_token = token_resp
            .id_token
            .ok_or_else(|| HappError::Invalid("missing id_token".to_string()))?;

        // Fetch JWKS and validate token signature
        let jwks = self.fetch_jwks().await?;
        let expected_iss = match self.cfg.tenant.as_str() {
            "common" | "organizations" | "consumers" => None,
            _ => Some(self.cfg.issuer()),
        };

        let verified = verify_entra_id_token(
            &id_token,
            &jwks,
            expected_iss.as_deref(),
            &self.cfg.client_id,
            &oidc.nonce,
        )
        .map_err(|e| HappError::Unauthorized(e))?;

        // Extract tid+oid
        let tid = verified
            .tid
            .clone()
            .ok_or_else(|| HappError::Invalid("missing tid".to_string()))?;
        let oid = verified
            .oid
            .clone()
            .ok_or_else(|| HappError::Invalid("missing oid".to_string()))?;

        if !policy.allowed_tenants.is_empty() && !policy.allowed_tenants.contains(&tid) {
            return Err(HappError::Unauthorized("tenant not allowed".to_string()));
        }

        if policy.require_mfa {
            let amr = verified.amr.clone().unwrap_or_default();
            if !amr.iter().any(|x| x == "mfa") {
                return Err(HappError::Unauthorized(
                    "MFA required but not present".to_string(),
                ));
            }
        }

        if !policy.required_auth_contexts.is_empty() {
            let acrs = verified.acrs.clone().unwrap_or_default();
            for required in policy.required_auth_contexts.iter() {
                if !acrs.iter().any(|a| a == required) {
                    return Err(HappError::Unauthorized(format!(
                        "required auth context {required} not satisfied"
                    )));
                }
            }
        }

        // auth_time freshness
        if let (Some(max_age), Some(auth_time)) = (policy.max_id_age_seconds, verified.auth_time) {
            let age = Utc::now().timestamp() - auth_time;
            if age > max_age as i64 {
                return Err(HappError::Unauthorized("identity too old".to_string()));
            }
        }

        // Build identity binding
        let embedded = policy.require_embedded_evidence;
        let evidence = IdentityEvidence {
            kind: "oidc_id_token".to_string(),
            token_hash: Some(format!(
                "sha256:{}",
                sha256_base64url_nopad(id_token.as_bytes())
            )),
            nonce_hash: Some(format!(
                "sha256:{}",
                sha256_base64url_nopad(oidc.nonce.as_bytes())
            )),
            embedded,
            token: if embedded {
                Some(id_token.clone())
            } else {
                None
            },
            jwks: if embedded { Some(jwks.clone()) } else { None },
        };

        let assurance = IdentityAssurance {
            auth_time: verified
                .auth_time
                .map(|t| Utc.timestamp_opt(t, 0).single())
                .flatten(),
            amr: verified.amr.unwrap_or_default(),
            acrs: verified.acrs.unwrap_or_default(),
        };

        let binding = IdentityBinding {
            mode: "verified".to_string(),
            scheme: "entra_oidc".to_string(),
            idp: Some(IdentityIdp {
                issuer: self.cfg.issuer(),
                tenant_id: Some(tid.clone()),
            }),
            subject: IdentitySubject {
                subject_type: "entra_oid_tid".to_string(),
                tid: Some(tid),
                oid: Some(oid),
                sub: verified.sub,
                extra: Default::default(),
            },
            assurance: Some(assurance),
            evidence,
        };

        provider.set_identity(&sid, binding)?;

        // Redirect back to the session page.
        Ok(IdentityAdapterOutcome::Completed)
    }
}

/// Minimal verified fields from an Entra ID token.
#[derive(Debug, Clone)]
struct VerifiedEntraToken {
    sub: Option<String>,
    tid: Option<String>,
    oid: Option<String>,
    amr: Option<Vec<String>>,
    acrs: Option<Vec<String>>,
    auth_time: Option<i64>,
}

/// Verify an Entra ID token using a provided JWKS.
/// This is intentionally minimal and should be hardened for production use.
fn verify_entra_id_token(
    token: &str,
    jwks: &Value,
    expected_iss: Option<&str>,
    expected_aud: &str,
    expected_nonce: &str,
) -> Result<VerifiedEntraToken, String> {
    use jsonwebtoken::jwk::{Jwk, JwkSet};
    use jsonwebtoken::{decode, decode_header, Algorithm, DecodingKey, Validation};

    let header = decode_header(token).map_err(|e| e.to_string())?;
    let kid = header.kid.ok_or_else(|| "missing kid".to_string())?;

    let jwk_set: JwkSet = serde_json::from_value(jwks.clone()).map_err(|e| e.to_string())?;
    let jwk: &Jwk = jwk_set
        .find(&kid)
        .ok_or_else(|| "kid not found in jwks".to_string())?;

    let decoding_key = DecodingKey::from_jwk(jwk).map_err(|e| e.to_string())?;

    let mut validation = Validation::new(Algorithm::RS256);
    validation.set_audience(&[expected_aud]);
    if let Some(iss) = expected_iss {
        validation.set_issuer(&[iss]);
    }
    validation.validate_exp = true;

    let td = decode::<serde_json::Value>(token, &decoding_key, &validation)
        .map_err(|e| e.to_string())?;
    let claims = td.claims;

    // nonce binding
    let nonce = claims
        .get("nonce")
        .and_then(|v| v.as_str())
        .ok_or_else(|| "missing nonce".to_string())?;
    if nonce != expected_nonce {
        return Err("nonce mismatch".to_string());
    }

    Ok(VerifiedEntraToken {
        sub: claims
            .get("sub")
            .and_then(|v| v.as_str())
            .map(|s| s.to_string()),
        tid: claims
            .get("tid")
            .and_then(|v| v.as_str())
            .map(|s| s.to_string()),
        oid: claims
            .get("oid")
            .and_then(|v| v.as_str())
            .map(|s| s.to_string()),
        amr: claims.get("amr").and_then(|v| v.as_array()).map(|arr| {
            arr.iter()
                .filter_map(|x| x.as_str().map(|s| s.to_string()))
                .collect()
        }),
        acrs: claims.get("acrs").and_then(|v| v.as_array()).map(|arr| {
            arr.iter()
                .filter_map(|x| x.as_str().map(|s| s.to_string()))
                .collect()
        }),
        auth_time: claims.get("auth_time").and_then(|v| v.as_i64()),
    })
}
